v3_async_wip/v3_async_wip/iothub_mqtt_client.py

```diff
@@ -79,40 +82,21 @@ def __init__(
         # Internal request/response infrastructure
         self._request_ledger = rr.RequestLedger()
         self._twin_responses_enabled = False
-        self._twin_response_listener = asyncio.create_task(self._process_twin_responses())
-
-    def _create_token_update_alarm(self) -> alarm.Alarm:
-        if not self._sastoken:
-            # This should never happen, it's just for the type checker
-            raise ValueError("Can't create alarm for no SASToken")
-
-        update_time = self._sastoken.expiry_time - DEFAULT_TOKEN_UPDATE_MARGIN
-
-        if isinstance(self._sastoken, st.RenewableSasToken):
-
-            def on_token_needs_update():
-                # Renew the token
-                logger.debug("Renewing SAS Token...")
-                try:
-                    self._sastoken.refresh()
-                    logger.debug("SAS Token renewal succeeded")
-                except st.SasTokenError:
-                    logger.error("SAS Token renewal failed")
-                    # TODO: background exception?
-                # With the token renewed, now set a new Alarm
-                self._sastoken_update_alarm = self._create_token_update_alarm()
 
+        # Background Tasks
```

Comment:
># Background Tasks [](http://example.com/codeflow?start=8&length=18)

I thought more about background exceptions. I don't think we need to invent our own mechanism for this. If we catch it and send it to our own "background exception handler", two things _don't_ happen:

1. Developers can't use "break on unhandled exceptions" in the debugger. Yes, if they know about the background exception handler, they can put code there, but what if they don't know about this? In V2, background exceptions were almost impossible to debug even if you knew about the background exception handler. All of the context was lost at that point -- it's way better to have the debugger break at the point of failure rather than inside some handler. 

2. `sys.execpthook` doesn't get called.
We had some github issue where the customer was talking about "uncatchable exceptions", and I think this is what they were talking about. I'd rather not have to explain why our exceptions are somehow different than all other exceptions because we have a special way for customers to find out about them.

I think the better way to architect things like this is to let the task fail and then recover afterwards. This way, the background exception is treated like all other exceptions, but we still have the benefit of being able to continue working.  Maybe `add_done_callback` can be used here?

Reply:
Yean we could probably do something like that. I'd have to explore it a bit more, this isn't something I've thought too much about yet, but resetting the task in a callback is an interesting idea.

That said, without piping them up, there's no way to actually catch them. A debugger could break in a background task, but we have no means to throw the exception without some kind of background exception handler. Then again, I'm really not sure what utility there really is in getting the exceptions directly to the user - what are they really supposed to do with that information dynamically at runtime? You can't realistically "respond" to them, because they didn't come from your inputs or actions. Either you injected bad code that is causing failure (that's not a dynamic change you can address) or you're receiving bad data from IoTHub/Edge/some other Module (again, not something you can address dynamically at runtime)